cardano-api/src/Cardano/Api/TxBody.hs

```diff
@@ -1054,6 +1073,274 @@ makeTransactionBody =
+          Mary.adHash' body ==
+          (Ledger.hashAuxiliaryData @StandardMary <$> mAux')
+
+
+fromLedgerTxIns
+  :: ShelleyBasedEra era
+  -> Ledger.TxBody (ShelleyLedgerEra era)
+  -> [(TxIn,BuildTxWith ViewTx (Witness WitCtxTxIn era))]
+fromLedgerTxIns era body =
+  [ (fromShelleyTxIn input, ViewTx)
+  | input <-
+      case era of
+        ShelleyBasedEraShelley -> toList $ Shelley._inputs body
+        ShelleyBasedEraAllegra -> toList $ Allegra.inputs' body
+        ShelleyBasedEraMary    -> toList $ Mary.inputs'    body
+  ]
+
+
+fromLedgerTxOuts
+  :: ShelleyBasedEra era -> Ledger.TxBody (ShelleyLedgerEra era) -> [TxOut era]
+fromLedgerTxOuts era body =
+  fromTxOut era <$>
+  case era of
+    ShelleyBasedEraShelley -> toList $ Shelley._outputs body
+    ShelleyBasedEraAllegra -> toList $ Allegra.outputs' body
+    ShelleyBasedEraMary    -> toList $ Mary.outputs'    body
+
+
+fromLedgerTxFee
+  :: ShelleyBasedEra era -> Ledger.TxBody (ShelleyLedgerEra era) -> TxFee era
+fromLedgerTxFee era body =
+  case era of
+    ShelleyBasedEraShelley ->
+      TxFeeExplicit TxFeesExplicitInShelleyEra $
+      fromShelleyLovelace $ Shelley._txfee body
+    ShelleyBasedEraAllegra ->
+      TxFeeExplicit TxFeesExplicitInAllegraEra $
+      fromShelleyLovelace $ Allegra.txfee' body
+    ShelleyBasedEraMary ->
+      TxFeeExplicit TxFeesExplicitInMaryEra $
+      fromShelleyLovelace $ Mary.txfee' body
+
+
+fromLedgerTxValidityRange
+  :: ShelleyBasedEra era
+  -> Ledger.TxBody (ShelleyLedgerEra era)
+  -> (TxValidityLowerBound era, TxValidityUpperBound era)
+fromLedgerTxValidityRange era body =
+  case era of
+    ShelleyBasedEraShelley ->
+      ( TxValidityNoLowerBound
+      , TxValidityUpperBound ValidityUpperBoundInShelleyEra $ Shelley._ttl body
+      )
+    ShelleyBasedEraAllegra ->
+      ( case invalidBefore of
+          SNothing -> TxValidityNoLowerBound
+          SJust s  -> TxValidityLowerBound ValidityLowerBoundInAllegraEra s
+      , case invalidHereafter of
+          SNothing -> TxValidityNoUpperBound ValidityNoUpperBoundInAllegraEra
+          SJust s  -> TxValidityUpperBound   ValidityUpperBoundInAllegraEra s
+      )
+      where
+        Allegra.ValidityInterval{invalidBefore, invalidHereafter} =
+          Allegra.vldt' body
+    ShelleyBasedEraMary ->
+      ( case invalidBefore of
+          SNothing -> TxValidityNoLowerBound
+          SJust s  -> TxValidityLowerBound ValidityLowerBoundInMaryEra s
+      , case invalidHereafter of
+          SNothing -> TxValidityNoUpperBound ValidityNoUpperBoundInMaryEra
+          SJust s  -> TxValidityUpperBound   ValidityUpperBoundInMaryEra s
+      )
+      where
+        Mary.ValidityInterval{invalidBefore, invalidHereafter} = Mary.vldt' body
+
+
+fromLedgerTxAuxiliaryData
+  :: ShelleyBasedEra era
+  -> Maybe (Ledger.AuxiliaryData (ShelleyLedgerEra era))
+  -> (TxMetadataInEra era, TxAuxScripts era)
+fromLedgerTxAuxiliaryData _ Nothing = (TxMetadataNone, TxAuxScriptsNone)
```

Comment:
This can be further simplified with:
```

fromLedgerAuxiliaryData
  :: ShelleyBasedEra era
  -> Ledger.AuxiliaryData (ShelleyLedgerEra era)
  -> (Map Word64 TxMetadataValue, [ScriptInEra era])
fromLedgerAuxiliaryData sbe aux =
  case sbe of
    ShelleyBasedEraShelley ->
      let Shelley.Metadata md = aux
      in (fromShelleyMetadata md, [])
    ShelleyBasedEraAllegra ->
      let Allegra.AuxiliaryData md sc = aux
      in (fromShelleyMetadata md, fromShelleyBasedScript sbe <$> toList sc)
    ShelleyBasedEraMary ->
      let Mary.AuxiliaryData md sc = aux
      in (fromShelleyMetadata md, fromShelleyBasedScript sbe <$> toList sc)
  ```

Reply:
We will be able to remove: `fromShelleyAuxiliaryData` and `fromAllegraBasedAuxiliaryData`